aten/src/ATen/Quantizer.h

```diff
@@ -0,0 +1,199 @@
+ * the instance and we can use it to quantize a float Tensor or dequantize a
+ * quantized Tensor.
+ *
+ * When you add new types of leaf Quantizer class, please also
+ * make sure to add a corresponding QScheme enum since
+ * they should have one to one mapping.
+ */
+struct CAFFE2_API Quantizer {
+  QScheme qscheme_;
+  Quantizer() {}
+  Quantizer(QScheme qscheme) : qscheme_(qscheme) {}
+  virtual ~Quantizer();
+
+  virtual QScheme qscheme() {
+    return qscheme_;
+  }
+
+  /**
+   * quantize a float Tensor into a quantized Tensor.
+   */
+  virtual QTensor quantize(RealTensor t) = 0;
+
+  /**
+   * dequantize a quantized Tensor into a float Tensor.
+   */
+  virtual RealTensor dequantize(QTensor t) = 0;
+};
+
+/**
+ * UniformQuantizer is the parent class for all uniform quantizers.
+ * These quantization scheme will map float value uniformly to
+ * the quantized value. For example, affine quantizer is
+ * the most commonly used scheme in this category.
+ */
+struct CAFFE2_API UniformQuantizer : public Quantizer {
+  UniformQuantizer() {}
+  UniformQuantizer(QScheme qscheme): Quantizer(qscheme) {}
+  virtual ~UniformQuantizer();
+};
+
+/**
+ * NonUniformQuantizer is the parent class for all non-uniform quantizers.
+ * These quantization scheme may map float value non-uniformly to the quantized
+ * value. K-means quantization is a representative example in this category.
+ */
+struct CAFFE2_API NonUniformQuantizer : public Quantizer {
+  NonUniformQuantizer() {}
+  NonUniformQuantizer(QScheme qscheme): Quantizer(qscheme) {}
+  virtual ~NonUniformQuantizer();
+
+};
+
+// There is also StochasticQuantizer which is uniform but not affine
+
+/**
+ * AffineQuantizer uses affine transformation to do quantization.
+ *
+ * For quantize:
+ * Y = clamp((X * scale + zero_point, min, max)
+ * For dequantize:
+ * X = (Y - zero_point) / scale
+ */
+struct CAFFE2_API AffineQuantizer : public UniformQuantizer {
+  AffineQuantizer() {}
+  AffineQuantizer(QScheme qscheme): UniformQuantizer(qscheme) {}
+  virtual ~AffineQuantizer();
+};
+
+/**
+ * SymmetricQuantizer is similar to AffineQuantizer except that it
+ * does not have zero_point
+ *
+ * For quantize:
+ * Y = clamp(X * scale, min, max)
+ * For dequantize:
+ * X = Y / scale
+ */
+struct CAFFE2_API SymmetricQuantizer : public UniformQuantizer {
+  SymmetricQuantizer() {}
+  SymmetricQuantizer(QScheme qscheme): UniformQuantizer(qscheme) {}
+  virtual ~SymmetricQuantizer();
+
+};
+
+/**
+ * PerLayerSymmetricQuantizer stores a single scale number which is
+ * used for quantizing all the values in the given Tensor
+ */
+struct CAFFE2_API PerLayerSymmetricQuantizer: public SymmetricQuantizer {
+  PerLayerSymmetricQuantizer() {}
+  PerLayerSymmetricQuantizer(float scale) : SymmetricQuantizer(kPerLayerSymmetric), scale_(scale) {}
+  virtual ~PerLayerSymmetricQuantizer();
+  float scale_{1.0};
+};
+
+/**
+ * PerChannelSymmetricQuantizer stores a vector of scale number and
```

Comment:
Maybe worth comment that per-channel quantization is mostly applied to output channels of weights? I don't think per-input channel of weight quantization or per-channel input quantization can be efficiently supported in most of processors because doing so would require each multiplication result within a single dot-product has a different scale.